sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -1122,7 +1119,7 @@ case class StringSpace(child: Expression)
   """)
 // scalastyle:on line.size.limit
 case class Substring(str: Expression, pos: Expression, len: Expression)
-  extends TernaryExpression with ImplicitCastInputTypes {
+  extends TernaryExpression with ImplicitCastInputTypes with NullIntolerant {
```

Comment:
At the beginning, when we introduce `NullIntolerant `, @marmbrus said we should do it more carefully. We prefer to using the white-list solution for avoiding hidden bugs. Marking it `NullIntolerant` if and only if we are ensure that they are null intolerant. Thus, when we doing it, we should also add the corresponding test cases and documents.